### PR TITLE
Adding a code action for expading classes

### PR DESCRIPTION
**Pull request by mamazu**
Just as discussed in #1021 and #1258 we should add a command to the language server that allows us to expand the current class name under the cursor.

### CODE REVIEW

File: lib/CodeTransform/Adapter/WorseReflection/Refactor/WorseExpandClass.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    public function getTextEdits(SourceCode $sourceCode, int $offset): TextDocumentEdits
+    {
+        $symbolInformation = $this->reflector
+            ->reflectOffset($sourceCode->__toString(), $offset)
+            ->symbolContext()
+            ->symbol();
+
+        $result = $this->classSearch->classSearch('filesystem', $symbolInformation->name());
```
Comment (mamazu) [inline lib/CodeTransform/Adapter/WorseReflection/Refactor/WorseExpandClass.php:40]: This is probably not the best idea. I think we should be able to look through the use statements and see if any class name there matches. What do you think?

File: lib/CodeTransform/Tests/Adapter/WorseReflection/Refactor/WorseExpandClassTest.php [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        [$source, $expected, $offset] = $this->sourceExpectedAndOffset(__DIR__ . '/fixtures/' . $test);
+
+        $this->classSeach->classSearch('filesystem', 'Foo')->willReturn([
+            [
+                'file_path' => '/home/mamazu/phpacator/phpactor/SomeNamespace/Foo.php',
+                'class' => 'SomeNamespace\\Foo',
+                'class_name' => 'Foo',
+                'class_namespace' => 'SomeNamespace',
+            ]
+        ])->shouldBeCalled();
```
Comment (mamazu) [inline lib/CodeTransform/Tests/Adapter/WorseReflection/Refactor/WorseExpandClassTest.php:37]: Is that even the correct kind of return value for this class?

File: lib/CodeTransform/Adapter/WorseReflection/Refactor/WorseExpandClass.php (hunk 1 above)
Comment (dantleech) [inline lib/CodeTransform/Adapter/WorseReflection/Refactor/WorseExpandClass.php:40]: no need to use "class search" and that class is very much legacy and should not be used here (note that it doesn't even come from a different "package" but actually from the "application".

you can just:

```
$type = $symbolInformation->type();
if (!$type instanceof ClassType) {
    return empty text edits;
}

$newClassName = $type->name()->__toString();
```

File: lib/CodeTransform/Tests/Adapter/WorseReflection/Refactor/WorseExpandClassTest.php (hunk 2 above)
Comment (dantleech) [reply lib/CodeTransform/Tests/Adapter/WorseReflection/Refactor/WorseExpandClassTest.php:37]: we should not use `classSearch`, you can do everything with the `Reflector` I think.

**Review by dantleech — COMMENTED** [review]
nice work! left some comments

File: lib/CodeTransform/Adapter/WorseReflection/Refactor/WorseExpandClass.php (hunk 1 above)
Comment (mamazu) [reply lib/CodeTransform/Adapter/WorseReflection/Refactor/WorseExpandClass.php]: Yeah, the new implementation looks much cleaner now as well.

**Comment by mamazu** [comment]
I replaced the class search with the code you suggested. Much more cleaner now.

File: lib/CodeTransform/Tests/Adapter/WorseReflection/Refactor/fixtures/expandClass1.test [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+<?php
+use SomeNamespace\Foo;
+
+echo 'Debugging ' . F<>oo::class;
+
+// File: expected
+<?php
+use SomeNamespace\Foo;
+
+echo 'Debugging ' . SomeNamespace\Foo::class;
```
Comment (dantleech) [inline lib/CodeTransform/Tests/Adapter/WorseReflection/Refactor/fixtures/expandClass1.test:11]: we need to add a namespaced test:

```
namespace Foo;
```

either in this case or in both cases the expanded class should be absolute `\SomeNamespace\Foo::class`

**Comment by dantleech** [comment]
hm. I've just realised that there is not really that useful. The original `:ExpandClass` would search for candidates and not require the class to already be imported. Expanding a class that's already imported isn't something you should ever need to do...

File: lib/CodeTransform/Tests/Adapter/WorseReflection/Refactor/fixtures/expandClass1.test (hunk 3 above)
Comment (mamazu) [reply lib/CodeTransform/Tests/Adapter/WorseReflection/Refactor/fixtures/expandClass1.test:11]: True

**Comment by mamazu** [comment]
Well, for classes that aren't imported we have the import command that does that. This would maybe help if you want to replace an import with a class reference directly like so:

```php
<?php
namespace Testing;

use Throwable;

try {
// something
} catch(Throwable) {}
```

You could convert that to:
```
<?php
namespace Testing;

try {
// something
} catch(\Throwable) {}
```

**Comment by dantleech** [comment]
Right, but you could also just type backslash. In all other cases it doesn't help. Class is already imported.

**Comment by dantleech** [comment]
The thing to consider is where does the value come in? I actually use `ExpandClass` quite frequently: but not in PHP files - I use it in XML and YAML files (e.g. service definitions). So I'd type `PaymentProcessor` and then expand it to `App\Payment\PaymentProcessor` or whatever.

Turning (imported) `(Foo\Bar)\Foo` to `\Foo\Bar\Foo`  in a PHP file doesn't really bring anything to the table.

**Comment by mamazu** [comment]
True that makes much more sense. But then it doesn't make sense to implement that in the language server with a php parser?

**Comment by dantleech** [comment]
In LS yes (needs to retrieve candidates to expand to) parser no....

**Comment by mamazu** [comment]
So what would be the best course of action? Rewrite the Code Action to be compatible with XML and Yaml? How would one go about that?

**Comment by dantleech** [comment]
We can certainly make a code action which acts the similar to the existing `:ExpandClass`. We can check the `$textDocument->extension() === 'xml'` for [example](https://github.com/phpactor/phpactor/blob/master/lib/Extension/Behat/ReferenceFinder/StepDefinitionLocator.php#L35), and use the `WordAtOffset` class to get the word at the given byte offset and use index search to get the class candidates and ask the user to select one if there is > 1 using the LSP client. Should it also work for PHP? Should it also "simply" the FQN?

But it' up to you, I don't think the goal is here is clearly defined. Would it make your day-to-day experience better in the editor?

**Comment by mamazu** [comment]
I mean I haven't really considered the use cases for this. For us we moved our configuration to php, which is much easier to deal with than XML and Yaml. But it also makes sense to implement it for XML and Yaml. 

I mean the best way would be to fully support XML configuration like in Symfonys case, have autocomplete for the places where classes are allowed.

**Comment by dantleech** [comment]
yeah, I don't really want to get into XML. Last time I checked it seemed I would have had to write an an XML parser for XML completion.

**Comment by mamazu** [comment]
True, and this would maybe also make the code action more complicated. But this would mean that if we can't use the PHP reflector that we would also have to revert to using the class search functionality?

**Comment by dantleech** [comment]
Well, we would use the index instead of whatever was used before, but yes, but that's necessary in anycsae (present a list of FQNs and choose one).

But I don't think it's clear what we're working or why atm. Is it an XML/HTML/Markdown/Blah plain text "fqn" expander (I'm not even sure that's a good idea, and if it were perhaps as an extension) or a PHP fqn expander (which as I said, has a limited use case)

**Comment by mamazu** [comment]
Well the idea was to replace the implementation of the `:ExpandClass` function which is currently implemented in VIM. If that makes sense or not that is still up for decision.

**Comment by dantleech** [comment]
What problem are we trying to solve - replacing an old command with a new one isn't solving a problem in itself. As said I find this useful primarily for expanding FQNs in XML and YAML (but also Markdown) and it doesn't really map to LSP as a code action as every single word would be eligible for expansion, and for PHP we have already established that it would bring little value. It would only be possible in it's current form as a custom Command which would then require a client-side addition (e.g. in the vscode client or as nvim configuration).

So short of opening a can of worms I would close this I think.

**Comment by mamazu** [comment]
Okay, I wanted to tackle those issues:

#1021 - This is directly mentioning what I was trying to implement. But it seems like that this doesn't make sense since the base functionality is already implemented. And implementing it as a code action wouldn't make sense (as by our discussion).

#1258 - The complete opposite. That would probably make the most sense as a code action. To be able to simplify the FQN to a relative name and add an import for it.

**Comment by mamazu** [comment]
Seeing that those two tickets have the complete opposite things in mind. I will just close the MR here and only to solve 1258, which I think is a cool functionality and could be useful when refactoring code.

**Comment by dantleech** [comment]
cool makes sense :+1:

**Pull request closed without merging**